### DISCUSSION TITLE
NullableAttribute Missing

### DISCUSSION THREAD

**Discussion by TaffarelJr** [cake-build/cake#4119] [Q&A]
I'm using [Cake.Tool 3.0.0](https://www.nuget.org/packages/Cake.Tool) to build my .NET 6/7 solution. I'm pulling in a few NuGet dependencies as Addins, including one combination Addin/Recipe package built in-house at my company. All are up-to-date versions.

However I'm having trouble debugging a recent error that popped up in the Cake scripts. Running Cake in `Diagnostic` mode, I can see:
- the .NET tools are all restored properly
- no modules are loaded (as expected)
- the in-house Recipe scripts are found and loaded
- the in-house Addin components are found and loaded
- assembly references all look good
- imported namespaces all look good

But when it gets to the step where Cake compiles the build script, it throws these two errors:

```console
Compiling build script...
(3472,115): error CS0234: The type or namespace name 'NullableAttribute' does not exist in the namespace 'System.Runtime.CompilerServices' (are you missing an assembly reference?)
(3472,115): error CS0234: The type or namespace name 'Nullable' does not exist in the namespace 'System.Runtime.CompilerServices' (are you missing an assembly reference?)   
Error: Cake.Core.CakeException: Error(s) occurred when compiling build script:
(3472,115): error CS0234: The type or namespace name 'NullableAttribute' does not exist in the namespace 'System.Runtime.CompilerServices' (are you missing an assembly reference?)
(3472,115): error CS0234: The type or namespace name 'Nullable' does not exist in the namespace 'System.Runtime.CompilerServices' (are you missing an assembly reference?)   
   at Cake.Infrastructure.Scripting.RoslynScriptSession.Execute(Script script) in C:\projects\cake\src\Cake\Infrastructure\Scripting\RoslynScriptSession.cs:line 186
   at Cake.Core.Scripting.ScriptRunner.Run(IScriptHost host, FilePath scriptPath) in C:\projects\cake\src\Cake.Core\Scripting\ScriptRunner.cs:line 172
   at Cake.Features.Building.BuildFeature.RunCore(IRemainingArguments arguments, BuildFeatureSettings settings) in C:\projects\cake\src\Cake\Features\Building\BuildFeature.cs:line 99
   at Cake.Features.Building.BuildFeature.Run(IRemainingArguments arguments, BuildFeatureSettings settings) in C:\projects\cake\src\Cake\Features\Building\BuildFeature.cs:line 49
   at Cake.Commands.DefaultCommand.Execute(CommandContext context, DefaultCommandSettings settings) in C:\projects\cake\src\Cake\Commands\DefaultCommand.cs:line 73
```

What I can infer is that the [`NullableAttribute`](https://github.com/dotnet/roslyn/blob/main/docs/features/nullable-metadata.md) class that is _supposed_ to be generated by the compiler, is not being generated. I can inspect each of the NuGet Addins and see that attribute exists in each of them, but for some reason Cake is not finding it -- and I really don't know how to go about debugging this further.

Is there a way to view/debug the actual Roslyn script created by Cake?

**Comment by devlead** [comment]
Does adding nullable directive help?

`#nullable enable`

**Comment by TaffarelJr** [comment]
At the top of one of the Recipe scripts, I added `#nullable enable` -- which made me add a few `?` and `!` operators to make everything work again. So the nullable reference types appear to be working. But I still get those 2 errors.

Just tried again with `#nullable enable` in the root `build.cake` file, just to see if the location matters. Same errors.

**Comment by TaffarelJr** [comment]
Additional information: I've stripped the whole project down to bare bones: all code commented out in the Cake Recipe scripts, the Cake Addin code, and the `build.cake` file; and now Cake loads the package, compiles, & runs properly (although it doesn't actually _do_ anything now).

I then started un-commenting one piece at a time, to see what actually causes the exeptions. I can see that Cake is loading the Recipes/Addin package properly (even when empty). I've added some small code back into the Recipe scripts and the Addin assembly, and they still work fine. So far so good.

> Side note: I did notice my assembly-level `[CakeNamespaceImport]` attributes in the Addin assembly aren't loaded into Cake unless there's an actual Alias method defined in that assembly. Is this behavior intentional? If so, it should be documented somewhere ...

I then un-commented a simple Alias method in the Addin assembly, and Cake still found it, compiled, & ran. But as soon as I changed the Alias signature to include nullable reference types (`string?`), I get the exceptions about `'NullableAttribute' does not exist in the namespace 'System.Runtime.CompilerServices'`. So it appears that is the cause: _if an Addin assembly contains an Alias method with a nullable reference type anywhere in the signature, Cake will still load the Addin assembly correctly--but will not generate the internal `NullableAttribute` definition needed for those nullable reference types._

Is there anything special I have to do to use nullable reference types in the Alias signature inside my Addin assembly?

**Comment by TaffarelJr** [comment]
Looking through the [Cake source code](https://github.com/cake-build/cake/blob/b7640388bc73a56dba048ade1575e435a0e83152/src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs#L131), it would seem the Roslyn compiler is responsible for finding/generating that attribute. (At least, in my limited knowledge, I can't see where Cake is to blame here.) I started a new [discussion in the Roslyn repo](https://github.com/dotnet/roslyn/discussions/67044) to see if they might have any answers, because this is a real blocker for me.

**Comment by TaffarelJr** [comment]
Alright, I've stripped my project down to the bare minimum that still reproduces the errors: [Cake.zip](https://github.com/cake-build/cake/files/11022112/Cake.zip)

It's a repo that includes a Cake Addin project. The goal is to compile the Cake Addin first (when `build.ps1` is executed), and then load it into the `build.cake` script so it's available for use. This part works fine (from what I can tell).

The problem comes once we get inside the Cake Addin code. It would appear that I can call a Cake Alias method that uses nullable reference types, no problem. BUT--if that Alias method calls _another_ Alias method that _also_ uses nullable reference types, then the Cake/Roslyn compiler bombs:
![image](https://user-images.githubusercontent.com/5290662/226442745-049ace8f-4ed8-4e03-8c61-c1c140f9b9b2.png)

**Reply by devlead** [reply]
I suspect it's a codegen issue and we should have some handling around nullable reference types.

**Reply by TaffarelJr** [reply]
Any idea if there's a workaround? Or if it's possible to estimate what might be involved, or how long it might take? As I mentioned before, this is a show-stopper for me at the moment ...
